tools/cmd/gomodcheck/main.go

```diff
@@ -0,0 +1,119 @@
+package main
```

Comment:
I think I would move this to hack/tools/gomodcheck
(as it's only meant for us and not for other folks writing controllers)

Reply:
Could we instead check the output from `go mod list` to achieve the same?

Reply:
example: `go mod graph | grep "^k8s.io/apiserver"` returns:

```
k8s.io/apiserver@v0.30.0-rc.1 github.com/coreos/go-oidc@v2.2.1+incompatible
k8s.io/apiserver@v0.30.0-rc.1 github.com/coreos/go-systemd/v22@v22.5.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/emicklei/go-restful/v3@v3.11.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/evanphx/json-patch@v4.12.0+incompatible
k8s.io/apiserver@v0.30.0-rc.1 github.com/fsnotify/fsnotify@v1.7.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/gogo/protobuf@v1.3.2
k8s.io/apiserver@v0.30.0-rc.1 github.com/google/cel-go@v0.17.8
k8s.io/apiserver@v0.30.0-rc.1 github.com/google/gnostic-models@v0.6.8
k8s.io/apiserver@v0.30.0-rc.1 github.com/google/go-cmp@v0.6.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/google/gofuzz@v1.2.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/google/uuid@v1.3.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/gorilla/websocket@v1.5.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/munnerz/goautoneg@v0.0.0-20191010083416-a7dc8b61c822
k8s.io/apiserver@v0.30.0-rc.1 github.com/mxk/go-flowrate@v0.0.0-20140419014527-cca7078d478f
k8s.io/apiserver@v0.30.0-rc.1 github.com/spf13/pflag@v1.0.5
k8s.io/apiserver@v0.30.0-rc.1 github.com/stretchr/testify@v1.8.4
k8s.io/apiserver@v0.30.0-rc.1 go.etcd.io/etcd/api/v3@v3.5.10
k8s.io/apiserver@v0.30.0-rc.1 go.etcd.io/etcd/client/pkg/v3@v3.5.10
k8s.io/apiserver@v0.30.0-rc.1 go.etcd.io/etcd/client/v3@v3.5.10
k8s.io/apiserver@v0.30.0-rc.1 go.etcd.io/etcd/server/v3@v3.5.10
k8s.io/apiserver@v0.30.0-rc.1 go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc@v0.42.0
k8s.io/apiserver@v0.30.0-rc.1 go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.44.0
k8s.io/apiserver@v0.30.0-rc.1 go.opentelemetry.io/otel@v1.19.0
k8s.io/apiserver@v0.30.0-rc.1 go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc@v1.19.0
k8s.io/apiserver@v0.30.0-rc.1 go.opentelemetry.io/otel/sdk@v1.19.0
k8s.io/apiserver@v0.30.0-rc.1 go.opentelemetry.io/otel/trace@v1.19.0
k8s.io/apiserver@v0.30.0-rc.1 go.uber.org/zap@v1.26.0
k8s.io/apiserver@v0.30.0-rc.1 golang.org/x/crypto@v0.19.0
k8s.io/apiserver@v0.30.0-rc.1 golang.org/x/net@v0.21.0
k8s.io/apiserver@v0.30.0-rc.1 golang.org/x/sync@v0.6.0
k8s.io/apiserver@v0.30.0-rc.1 golang.org/x/sys@v0.17.0
k8s.io/apiserver@v0.30.0-rc.1 golang.org/x/time@v0.3.0
k8s.io/apiserver@v0.30.0-rc.1 google.golang.org/genproto/googleapis/api@v0.0.0-20230726155614-23370e0ffb3e
k8s.io/apiserver@v0.30.0-rc.1 google.golang.org/grpc@v1.58.3
k8s.io/apiserver@v0.30.0-rc.1 google.golang.org/protobuf@v1.33.0
k8s.io/apiserver@v0.30.0-rc.1 gopkg.in/natefinch/lumberjack.v2@v2.2.1
k8s.io/apiserver@v0.30.0-rc.1 gopkg.in/square/go-jose.v2@v2.6.0
k8s.io/apiserver@v0.30.0-rc.1 k8s.io/api@v0.30.0-rc.1
k8s.io/apiserver@v0.30.0-rc.1 k8s.io/apimachinery@v0.30.0-rc.1
k8s.io/apiserver@v0.30.0-rc.1 k8s.io/client-go@v0.30.0-rc.1
k8s.io/apiserver@v0.30.0-rc.1 k8s.io/component-base@v0.30.0-rc.1
k8s.io/apiserver@v0.30.0-rc.1 k8s.io/klog/v2@v2.120.1
k8s.io/apiserver@v0.30.0-rc.1 k8s.io/kms@v0.30.0-rc.1
k8s.io/apiserver@v0.30.0-rc.1 k8s.io/kube-openapi@v0.0.0-20240228011516-70dd3763d340
k8s.io/apiserver@v0.30.0-rc.1 k8s.io/utils@v0.0.0-20230726121419-3b25d923346b
k8s.io/apiserver@v0.30.0-rc.1 sigs.k8s.io/apiserver-network-proxy/konnectivity-client@v0.29.0
k8s.io/apiserver@v0.30.0-rc.1 sigs.k8s.io/json@v0.0.0-20221116044647-bc3834ca7abd
k8s.io/apiserver@v0.30.0-rc.1 sigs.k8s.io/structured-merge-diff/v4@v4.4.1
k8s.io/apiserver@v0.30.0-rc.1 sigs.k8s.io/yaml@v1.3.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/NYTimes/gziphandler@v1.1.1
k8s.io/apiserver@v0.30.0-rc.1 github.com/antlr/antlr4/runtime/Go/antlr/v4@v4.0.0-20230305170008-8188dc5388df
k8s.io/apiserver@v0.30.0-rc.1 github.com/asaskevich/govalidator@v0.0.0-20190424111038-f61b66f89f4a
k8s.io/apiserver@v0.30.0-rc.1 github.com/beorn7/perks@v1.0.1
k8s.io/apiserver@v0.30.0-rc.1 github.com/blang/semver/v4@v4.0.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/cenkalti/backoff/v4@v4.2.1
k8s.io/apiserver@v0.30.0-rc.1 github.com/cespare/xxhash/v2@v2.2.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/coreos/go-semver@v0.3.1
k8s.io/apiserver@v0.30.0-rc.1 github.com/davecgh/go-spew@v1.1.1
k8s.io/apiserver@v0.30.0-rc.1 github.com/dustin/go-humanize@v1.0.1
k8s.io/apiserver@v0.30.0-rc.1 github.com/felixge/httpsnoop@v1.0.3
k8s.io/apiserver@v0.30.0-rc.1 github.com/go-logr/logr@v1.4.1
k8s.io/apiserver@v0.30.0-rc.1 github.com/go-logr/stdr@v1.2.2
k8s.io/apiserver@v0.30.0-rc.1 github.com/go-openapi/jsonpointer@v0.19.6
k8s.io/apiserver@v0.30.0-rc.1 github.com/go-openapi/jsonreference@v0.20.2
k8s.io/apiserver@v0.30.0-rc.1 github.com/go-openapi/swag@v0.22.3
k8s.io/apiserver@v0.30.0-rc.1 github.com/golang-jwt/jwt/v4@v4.5.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/golang/groupcache@v0.0.0-20210331224755-41bb18bfe9da
k8s.io/apiserver@v0.30.0-rc.1 github.com/golang/protobuf@v1.5.4
k8s.io/apiserver@v0.30.0-rc.1 github.com/google/btree@v1.0.1
k8s.io/apiserver@v0.30.0-rc.1 github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/grpc-ecosystem/grpc-gateway@v1.16.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/grpc-ecosystem/grpc-gateway/v2@v2.16.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/imdario/mergo@v0.3.6
k8s.io/apiserver@v0.30.0-rc.1 github.com/inconshreveable/mousetrap@v1.1.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/jonboulle/clockwork@v0.2.2
k8s.io/apiserver@v0.30.0-rc.1 github.com/josharian/intern@v1.0.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/json-iterator/go@v1.1.12
k8s.io/apiserver@v0.30.0-rc.1 github.com/mailru/easyjson@v0.7.7
k8s.io/apiserver@v0.30.0-rc.1 github.com/matttproud/golang_protobuf_extensions@v1.0.4
k8s.io/apiserver@v0.30.0-rc.1 github.com/moby/spdystream@v0.2.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/modern-go/concurrent@v0.0.0-20180306012644-bacd9c7ef1dd
k8s.io/apiserver@v0.30.0-rc.1 github.com/modern-go/reflect2@v1.0.2
k8s.io/apiserver@v0.30.0-rc.1 github.com/pkg/errors@v0.9.1
k8s.io/apiserver@v0.30.0-rc.1 github.com/pmezard/go-difflib@v1.0.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/pquerna/cachecontrol@v0.1.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/prometheus/client_golang@v1.16.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/prometheus/client_model@v0.4.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/prometheus/common@v0.44.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/prometheus/procfs@v0.10.1
k8s.io/apiserver@v0.30.0-rc.1 github.com/sirupsen/logrus@v1.9.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/soheilhy/cmux@v0.1.5
k8s.io/apiserver@v0.30.0-rc.1 github.com/spf13/cobra@v1.7.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/stoewer/go-strcase@v1.2.0
k8s.io/apiserver@v0.30.0-rc.1 github.com/tmc/grpc-websocket-proxy@v0.0.0-20220101234140-673ab2c3ae75
k8s.io/apiserver@v0.30.0-rc.1 github.com/xiang90/probing@v0.0.0-20190116061207-43a291ad63a2
k8s.io/apiserver@v0.30.0-rc.1 go.etcd.io/bbolt@v1.3.8
k8s.io/apiserver@v0.30.0-rc.1 go.etcd.io/etcd/client/v2@v2.305.10
k8s.io/apiserver@v0.30.0-rc.1 go.etcd.io/etcd/pkg/v3@v3.5.10
k8s.io/apiserver@v0.30.0-rc.1 go.etcd.io/etcd/raft/v3@v3.5.10
k8s.io/apiserver@v0.30.0-rc.1 go.opentelemetry.io/otel/exporters/otlp/otlptrace@v1.19.0
k8s.io/apiserver@v0.30.0-rc.1 go.opentelemetry.io/otel/metric@v1.19.0
k8s.io/apiserver@v0.30.0-rc.1 go.opentelemetry.io/proto/otlp@v1.0.0
k8s.io/apiserver@v0.30.0-rc.1 go.uber.org/multierr@v1.11.0
k8s.io/apiserver@v0.30.0-rc.1 golang.org/x/exp@v0.0.0-20220722155223-a9213eeb770e
k8s.io/apiserver@v0.30.0-rc.1 golang.org/x/oauth2@v0.10.0
k8s.io/apiserver@v0.30.0-rc.1 golang.org/x/term@v0.17.0
k8s.io/apiserver@v0.30.0-rc.1 golang.org/x/text@v0.14.0
k8s.io/apiserver@v0.30.0-rc.1 google.golang.org/appengine@v1.6.7
k8s.io/apiserver@v0.30.0-rc.1 google.golang.org/genproto@v0.0.0-20230803162519-f966b187b2e5
k8s.io/apiserver@v0.30.0-rc.1 google.golang.org/genproto/googleapis/rpc@v0.0.0-20230822172742-b8732ec3820d
k8s.io/apiserver@v0.30.0-rc.1 gopkg.in/inf.v0@v0.9.1
k8s.io/apiserver@v0.30.0-rc.1 gopkg.in/yaml.v2@v2.4.0
k8s.io/apiserver@v0.30.0-rc.1 gopkg.in/yaml.v3@v3.0.1
k8s.io/apiserver@v0.30.0-rc.1 go@1.22.0
```

Reply:
It's a more elegant solution, I'll proceed with that change

Reply:
@vincepri I made the tool able to verify modules onto any upstream using your suggestion.